### PR TITLE
back up user interactions with jubSignal

### LINKED ISSUE #57: Support outbound tap in jubsignal

### PR DESCRIPTION
**Pull request by AndrewCLu**
- [x] Define data types for different jubSignal message types - outbound taps, inbound taps, location taps, and quests completed
- [x] Send yourself a jubSignal message when tapping another user
- [x] Add support for private notes when tapping another user
- [x] Send yourself a jubSignal message when tapping a location card  

Closes #57 #55

### CODE REVIEW

File: src/components/LoginForm.tsx [hunk 1]
```diff
@@ -28,12 +47,219 @@ export default function LoginForm({
...
+          case JUB_SIGNAL_MESSAGE_TYPE.QUEST_COMPLETED:
+            throw new Error("Unable to handle quest completed messages");
+          default:
+            throw new Error("Invalid message type");
+        }
+      } catch (error) {
+        console.error("Invalid message received from server: ", message);
+      }
+    });
+
```
Comment (RiverRuby) [inline src/components/LoginForm.tsx:251]: In general, this entire section doesn't really have any real error catching. Just console errors. We should have some sort of error screen to send users to, as well as info on who to contact.

File: src/lib/client/localStorage/users.ts [hunk 2]
```diff
@@ -9,10 +9,16 @@ import {
 export const USERS_STORAGE_KEY = "users";
 
 export type User = {
-  displayName: string;
-  encryptionPublicKey: string;
-  twitterUsername?: string;
-  telegramUsername?: string;
+  name: string; // User's display name
+  encPk: string; // User's encryption public key
+  x?: string; // User's Twitter username
```
Comment (RiverRuby) [inline src/lib/client/localStorage/users.ts:14]: Slightly confusing variable name for twitter

**Review by RiverRuby — APPROVED** [review]
Overall functionality looks good to me! Going to test more rigorously in next PR.

**Review by RiverRuby — APPROVED** [review]
Overall functionality looks good to me! Going to test more rigorously in next PR. There are some comments re: error messages but we can just make a ticket for that later.

File: src/components/LoginForm.tsx [hunk 3]
```diff
@@ -28,12 +47,219 @@ export default function LoginForm({
...
+    if (!response.ok) {
+      console.error("Error fetching jubSignal messages from server");
+      onFailedLogin("Error logging in. Please try again.");
+      return;
+    }
+    const messages = await response.json();
+
+    // Load backup into localStorage and fetch user profile and keys
+    const { profile, keys } = loadBackup(backup);
```
[the remark this answers is not in the document]
Comment (AndrewCLu) [reply src/components/LoginForm.tsx:87]: Added in https://github.com/nfc-denver/nfc-denver/pull/64

File: src/pages/api/messages.ts [hunk 4]
```diff
@@ -31,17 +31,19 @@ export default async function handler(
...
+      dateFilter["lte"] = new Date(endDate);
     }
 
     const receivedMessages = await prisma.message.findMany({
       where: {
         recipientId: userId,
-        createdAt: {
-          gte: new Date(startDate),
-        },
+        createdAt: dateFilter,
```
[the remark this answers is not in the document]
Comment (AndrewCLu) [reply src/pages/api/messages.ts:46]: Yes, works if we only have one filter or another. Will double check this when adding refresh logic

File: src/components/LoginForm.tsx (hunk 1 above)
Comment (AndrewCLu) [reply src/components/LoginForm.tsx:251]: This applies to pretty much all the flows we have now - adding a ticket here: https://github.com/nfc-denver/nfc-denver/issues/67

File: src/lib/client/localStorage/users.ts (hunk 2 above)
Comment (AndrewCLu) [reply src/lib/client/localStorage/users.ts:14]: Its the shortest one that makes sense

**Pull request merged**